events/events.c

```diff
@@ -189,6 +201,33 @@ events_spin(int * done)
 	return (rc);
 }
 
+/**
+ * events_interrupt():
+ * Halt the event loop after finishing the current event.  This function can
+ * be safely called from within a signal handler.
+ */
+void
+events_interrupt(void)
+{
+
+	/* Interrupt the event loop. */
+	interrupt_requested = 1;
+}
+
+/**
+ * events_interrupt_cancel():
+ * Cancel a requested interrupt from events_interrupt.  Must be used before
+ * running another event loop.  This function can be safely called from within
+ * a signal handler.
+ */
+void
+events_interrupt_cancel(void)
```

Comment:
API question: I considered automatically resetting `interrupt_requested = 0` once it was used, but then I figured that explicit was better than implicit, so I added this function instead.  Do you want application code to call `events_interrupt_cancel()` before running `events_run()` again?

(I also considered automatically setting `interrupt_requested=0` at the top of `events_run()` and `events_spin()`, but since the latter calls the former, that could end up with an interrupt being ignored.)

Reply:
Good question.  It should be reset automatically before `events_spin` and `events_run` return; otherwise we could "lose a wakeup" if a signal arrives at the wrong time.

Let's have the existing `events_run` become an internal function and have wrappers around it.